runner/src/internal.rs

```diff
@@ -97,24 +97,76 @@ pub struct BuildServer {
     pub coverage: bool,
 }
 
+#[derive(Clone)]
+/// A construct to keep track of the status of the execution. It only cares about the top-level
+/// steps.
+pub struct Status {
+    failed: usize,
+    passed: usize,
+    remaining: usize,
+}
+
+impl Status {
+    pub fn new(remaining: usize) -> Self {
+        Status {
+            failed: 0,
+            passed: 0,
+            remaining,
+        }
+    }
+
+    /// Guarantees that the `failed`, `passed`, and `remaining` counts are updated only after the
+    /// completions of each top-level step.
+    fn update(&mut self, context: &Context, step_failed: bool) {
+        // Update the status with results from the step, only if it is a top-level step.
+        if context.depth == 1 {
+            self.remaining -= 1;
+            // We only care about pass and fail. If an entire step is skipped, we count it as a
+            // passed step.
+            if step_failed {
+                self.failed += 1;
+            } else {
+                self.passed += 1;
+            }
+        }
+    }
+}
+
+/// Formats the status as `F:<failed-count>,P:<passed_count>,R:<remaining_count>`, suitable for
+/// annotating the log lines.
+impl std::fmt::Display for Status {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(
+            f,
+            "F:{},P:{},R:{}",
+            self.failed, self.passed, self.remaining
+        )
+    }
+}
+
 /// Encapsulates all the local state relative to a step, and is propagated to child steps.
 pub struct Context {
     opt: Opt,
     prefix: String,
+    /// For convenience; to make updating of the `Status` easier.
+    /// It matches the number of atomic step names in the `prefix`.
+    depth: u8,
```

Comment:
May be simpler / more elegant to change prefix to a `Vec<String>` with the various segments, and just count them.

Reply:
I like that idea, it seems nice :)